File: crates/microvecdb-python/python/minivecdb/cache.py
"""
minivecdb.cache — Semantic cache for LLM responses.

Stores prompt→response pairs as 384-dim embeddings. On each lookup it
finds the most similar cached prompt; if the cosine similarity exceeds
``similarity_threshold`` the stored response is returned instantly.

LangChain integration
---------------------
``MiniVecDbSemanticCache`` extends ``langchain_core.caches.BaseCache``
so it can be set as the global LangChain cache::

    import langchain
    from minivecdb.cache import MiniVecDbSemanticCache
    from langchain_openai import OpenAIEmbeddings

    langchain.llm_cache = MiniVecDbSemanticCache(
        embedding_function=OpenAIEmbeddings(
            model="text-embedding-3-small", dimensions=384
        ),
        similarity_threshold=0.92,
        ttl_minutes=1440,  # 24 h
    )

Standalone usage
----------------
::

    from minivecdb.cache import MiniVecDbSemanticCache

    cache = MiniVecDbSemanticCache(embedding_function=my_embed_fn)
    cache.update("What are your opening hours?", "llm", [Generation(text="9am-6pm")])
    result = cache.lookup("When do you open?", "llm")  # → [Generation(text="9am-6pm")]
"""

from __future__ import annotations

import threading
import time
from typing import Any, Callable, Optional, Sequence, Union

from minivecdb import MiniVecDb

try:
    from langchain_core.caches import BaseCache
    from langchain_core.outputs import Generation

    RETURN_VAL_TYPE = Sequence[Generation]
    _HAS_LANGCHAIN = True
except ImportError:  # pragma: no cover
    BaseCache = object  # type: ignore[assignment,misc]
    Generation = None  # type: ignore[assignment]
    RETURN_VAL_TYPE = Any  # type: ignore[misc]
    _HAS_LANGCHAIN = False


# ── helpers ───────────────────────────────────────────────────────────────────

def _make_embed_fn(
    embedding_function: Union[str, Callable, Any],
) -> Callable[[str], list[float]]:
    """
    Accept:
    • A plain callable ``(str) -> list[float]``
    • A LangChain ``Embeddings`` instance (has ``.embed_query``)
    • A string model name (OpenAI shortcut — requires ``langchain-openai``)
    """
    if callable(embedding_function) and not hasattr(embedding_function, "embed_query"):
        return embedding_function

    if hasattr(embedding_function, "embed_query"):
        return embedding_function.embed_query

    if isinstance(embedding_function, str):
        try:
            from langchain_openai import OpenAIEmbeddings  # type: ignore[import]
        except ImportError as exc:  # pragma: no cover
            raise ImportError(
                "pip install langchain-openai to use a string model name, "
                "or pass an Embeddings instance / callable directly."
            ) from exc
        emb = OpenAIEmbeddings(model=embedding_function, dimensions=384)
        return emb.embed_query

    raise TypeError(
        "embedding_function must be a callable, a LangChain Embeddings instance, "
        f"or a model-name string. Got: {type(embedding_function)!r}"
    )


class _RepeatingTimer:
    """Daemon timer that re-schedules itself after each tick."""

    def __init__(self, interval_s: float, fn: Callable[[], None]) -> None:
        self._interval = interval_s
        self._fn = fn
        self._timer: Optional[threading.Timer] = None
        self._stopped = False

    def start(self) -> None:
        self._schedule()

    def _schedule(self) -> None:
        if self._stopped:
            return
        self._timer = threading.Timer(self._interval, self._run)
        self._timer.daemon = True
        self._timer.start()

    def _run(self) -> None:
        self._fn()
        self._schedule()

    def cancel(self) -> None:
        self._stopped = True
        if self._timer is not None:
            self._timer.cancel()


# ── MiniVecDbSemanticCache ────────────────────────────────────────────────────

class MiniVecDbSemanticCache(BaseCache):  # type: ignore[misc]
    """
    LangChain-compatible semantic cache backed by MiniVecDb.

    Parameters
    ----------
    embedding_function:
        How to embed prompts into 384-dim vectors. Accepts:

        * A callable ``(str) -> list[float]``
        * A LangChain ``Embeddings`` instance (e.g. ``OpenAIEmbeddings``)
        * A string model name, e.g. ``"text-embedding-3-small"``
          (requires ``pip install langchain-openai``)

    similarity_threshold:
        Cosine similarity above which a lookup counts as a hit.
        Default 0.95. Lower → more hits, less accuracy.

    ttl_minutes:
        How many minutes a cached response lives before being evicted.
        0 (default) = immortal.

    gc_interval_seconds:
        How often the background GC thread runs. Default 60 s.
        Only relevant when ``ttl_minutes > 0``.

    db_instance:
        Optional pre-created ``MiniVecDb`` instance. A new one is created
        automatically if omitted.
    """

    def __init__(
        self,
        embedding_function: Union[str, Callable[[str], list[float]], Any],
        *,
        similarity_threshold: float = 0.95,
        ttl_minutes: float = 0,
        gc_interval_seconds: float = 60,
        db_instance: Optional[MiniVecDb] = None,
    ) -> None:
        if not _HAS_LANGCHAIN:  # pragma: no cover
            raise ImportError(
                "pip install langchain-core to use MiniVecDbSemanticCache."
            )
        self._db: MiniVecDb = db_instance if db_instance is not None else MiniVecDb()
        self._embed: Callable[[str], list[float]] = _make_embed_fn(embedding_function)
        self._threshold = similarity_threshold
        self._ttl_ms = ttl_minutes * 60_000
        self._entries: dict[int, dict[str, Any]] = {}
        self._next_id = 0
        self._lock = threading.Lock()

        self._timer: Optional[_RepeatingTimer] = None
        if self._ttl_ms > 0:
            self._timer = _RepeatingTimer(gc_interval_seconds, self._gc_tick)
            self._timer.start()

    # ----- BaseCache interface -----------------------------------------------

    def lookup(self, prompt: str, llm_string: str) -> Optional[RETURN_VAL_TYPE]:
        """Return cached generations for *prompt* or ``None`` on a miss."""
        with self._lock:
            if not self._entries:
                return None
            vec = self._embed(prompt)
            results = self._db.search(vec, k=1)
            if not results:
                return None
            doc_id, score = results[0]
            if score < self._threshold:
                return None
            entry = self._entries.get(doc_id)
            return entry["response"] if entry is not None else None

    def update(
        self,
        prompt: str,
        llm_string: str,
        return_val: RETURN_VAL_TYPE,
    ) -> None:
        """Store *return_val* under the embedding of *prompt*."""
        with self._lock:
            vec = self._embed(prompt)
            doc_id = self._next_id
            self._next_id += 1
            self._db.insert(doc_id, vec)
            self._entries[doc_id] = {
                "response": return_val,
                "inserted_at_ms": time.time() * 1_000,
            }

    def clear(self, **kwargs: Any) -> None:
        """Wipe the entire cache."""
        with self._lock:
            self._db = MiniVecDb()
            self._entries.clear()
            self._next_id = 0

    # ----- Extras ------------------------------------------------------------

    def run_gc(self) -> int:
        """Manually run one GC cycle. Returns number of entries evicted."""
        if self._ttl_ms <= 0:
            return 0
        return self._gc_tick()

    def destroy(self) -> None:
        """Stop the GC timer. Call when the cache is no longer needed."""
        if self._timer is not None:
            self._timer.cancel()
            self._timer = None

    # ----- Private -----------------------------------------------------------

    def _gc_tick(self) -> int:
        with self._lock:
            tombstoned = self._db.run_gc(int(self._ttl_ms))
            cutoff_ms = time.time() * 1_000 - self._ttl_ms
            stale = [k for k, v in self._entries.items()
                     if v["inserted_at_ms"] < cutoff_ms]
            for k in stale:
                del self._entries[k]
            return tombstoned

File: packages/core/src/cache.ts
import { embed } from 'ai';
import type { EmbeddingModel } from 'ai';
import { MicroVecDB } from './MicroVecDB.js';
import type { DbOptions } from './types.js';

const WASM_DIMS = 384;

// ----- Public types ----------------------------------------------------------

export interface SemanticCacheOptions extends DbOptions {
  /**
   * Vercel AI SDK embedding model — must produce 384-dim vectors.
   * e.g. `openai.embedding('text-embedding-3-small', { dimensions: 384 })`
   */
  embeddingModel: EmbeddingModel;
  /**
   * Minimum cosine similarity to count as a cache hit. Default: 0.95.
   * Lower = more aggressive cache (more hits, lower accuracy).
   */
  threshold?: number;
  /**
   * How long cached responses stay alive, in minutes. 0 = immortal.
   */
  ttlMinutes?: number;
  /** GC interval in ms. Default: 60 000. Only relevant when ttlMinutes > 0. */
  gcIntervalMs?: number;
}

interface CacheEntry {
  response: string;
  insertedAt: number;
}

// ----- SemanticCache ---------------------------------------------------------

/**
 * Semantic cache for LLM responses backed by MicroVecDB.
 *
 * Stores prompt→response pairs as 384-dim embeddings. On each call it looks
 * for a semantically similar prompt already in cache; if the cosine similarity
 * exceeds `threshold` the stored response is returned instantly without any
 * LLM call.
 *
 * @example
 * ```ts
 * import { openai } from '@ai-sdk/openai';
 * import { generateText } from 'ai';
 * import { withSemanticCache } from '@microvecdb/core/cache';
 *
 * const cachedGenerate = await withSemanticCache(
 *   {
 *     embeddingModel: openai.embedding('text-embedding-3-small', { dimensions: 384 }),
 *     threshold: 0.95,
 *     ttlMinutes: 60,
 *   },
 *   async (prompt) => {
 *     const { text } = await generateText({ model: openai('gpt-4o'), prompt });
 *     return text;
 *   },
 * );
 *
 * // First call → real LLM, subsequent similar calls → instant cache hit.
 * const answer = await cachedGenerate(userQuestion);
 * ```
 */
export class SemanticCache {
  readonly #db: MicroVecDB;
  readonly #model: EmbeddingModel;
  readonly #threshold: number;
  readonly #ttlMs: number;
  readonly #entries = new Map<number, CacheEntry>();
  #nextId = 0;
  #gcTimer: ReturnType<typeof setInterval> | null = null;

  private constructor(db: MicroVecDB, opts: SemanticCacheOptions) {
    this.#db        = db;
    this.#model     = opts.embeddingModel;
    this.#threshold = opts.threshold   ?? 0.95;
    this.#ttlMs     = (opts.ttlMinutes ?? 0) * 60_000;

    if (this.#ttlMs > 0) {
      const interval = opts.gcIntervalMs ?? 60_000;
      this.#gcTimer = setInterval(() => { this.#gcTick(); }, interval);
      if (
        this.#gcTimer !== null &&
        typeof this.#gcTimer === 'object' &&
        'unref' in this.#gcTimer
      ) {
        (this.#gcTimer as NodeJS.Timeout).unref();
      }
    }
  }

  // ----- Factory ------------------------------------------------------------

  static async create(opts: SemanticCacheOptions): Promise<SemanticCache> {
    const db = await MicroVecDB.init(opts);
    return new SemanticCache(db, opts);
  }

  // ----- Core operations ----------------------------------------------------

  /**
   * Look up a prompt in the cache.
   * Returns the cached response string, or `null` on a miss.
   */
  async lookup(prompt: string): Promise<string | null> {
    if (this.#entries.size === 0) return null;

    const { embedding } = await embed({ model: this.#model, value: prompt });
    this.#assertDims(embedding.length);

    const hits = this.#db.search(new Float32Array(embedding), { limit: 1 });
    if (hits.length === 0) return null;

    const [hit] = hits;
    if (hit.score < this.#threshold) return null;

    return this.#entries.get(hit.id)?.response ?? null;
  }

  /**
   * Store a prompt→response pair in the cache.
   */
  async set(prompt: string, response: string): Promise<void> {
    const { embedding } = await embed({ model: this.#model, value: prompt });
    this.#assertDims(embedding.length);

    const id = this.#nextId++;
    this.#db.insert({ id, vector: new Float32Array(embedding) });
    this.#entries.set(id, { response, insertedAt: Date.now() });
  }

  /**
   * Wrap a fallback function with cache-aside logic.
   *
   * The returned function:
   *   1. Checks the cache for a semantically similar prompt.
   *   2. On hit: returns the cached response.
   *   3. On miss: calls `fn`, stores the result, then returns it.
   */
  wrap(fn: (prompt: string) => Promise<string>): (prompt: string) => Promise<string> {
    return async (prompt: string): Promise<string> => {
      const cached = await this.lookup(prompt);
      if (cached !== null) return cached;
      const response = await fn(prompt);
      await this.set(prompt, response);
      return response;
    };
  }

  // ----- TTL / GC -----------------------------------------------------------

  /** Run one GC cycle manually. Returns the number of entries evicted. */
  runGc(): number {
    if (this.#ttlMs <= 0) return 0;
    return this.#gcTick();
  }

  // ----- Lifecycle ----------------------------------------------------------

  /** Stop the GC timer and release WASM memory. Safe to call multiple times. */
  destroy(): void {
    if (this.#gcTimer !== null) {
      clearInterval(this.#gcTimer);
      this.#gcTimer = null;
    }
    this.#entries.clear();
    this.#db.dispose();
  }

  // ----- Private ------------------------------------------------------------

  #assertDims(got: number): void {
    if (got !== WASM_DIMS) {
      throw new RangeError(
        `[SemanticCache] expected ${WASM_DIMS}-dim embedding, got ${got}. ` +
        `Pass a 384-dim model, e.g. openai.embedding('text-embedding-3-small', { dimensions: 384 }).`,
      );
    }
  }

  #gcTick(): number {
    const tombstoned = this.#db.runGc(this.#ttlMs);
    const cutoff = Date.now() - this.#ttlMs;
    for (const [id, entry] of this.#entries) {
      if (entry.insertedAt < cutoff) this.#entries.delete(id);
    }
    return tombstoned;
  }
}

// ----- Convenience HOF -------------------------------------------------------

/**
 * Wrap an async function that calls an LLM with semantic caching.
 *
 * Identical prompts — or semantically very similar ones — skip the LLM call
 * entirely and return the cached response. The cache self-expires after
 * `ttlMinutes` minutes (0 = never).
 *
 * @example
 * ```ts
 * const cachedGenerate = await withSemanticCache(
 *   {
 *     embeddingModel: openai.embedding('text-embedding-3-small', { dimensions: 384 }),
 *     threshold: 0.95,
 *     ttlMinutes: 60,
 *   },
 *   async (prompt) => {
 *     const { text } = await generateText({ model: openai('gpt-4o'), prompt });
 *     return text;
 *   },
 * );
 *
 * const answer = await cachedGenerate(userQuestion);
 * ```
 */
export async function withSemanticCache(
  opts: SemanticCacheOptions,
  fallback: (prompt: string) => Promise<string>,
): Promise<(prompt: string) => Promise<string>> {
  const cache = await SemanticCache.create(opts);
  return cache.wrap(fallback);
}
